packages/expo-cellular/android/src/main/java/expo/modules/cellular/CellularModule.kt

```diff
@@ -0,0 +1,152 @@
+    constants["isoCountryCode"] = telephonyManager.simCountryIso
+
+    //check if sim state is ready
+    if (telephonyManager.simState == TelephonyManager.SIM_STATE_READY) {
+      constants["carrier"] = telephonyManager.simOperatorName
+      constants["mobileCountryCode"] = telephonyManager.simOperator.substring(0, 3)
+      constants["mobileNetworkCode"] = StringBuilder(telephonyManager.simOperator).delete(0, 3).toString()
+    } else {
+      constants["carrier"] = null
+      constants["mobileCountryCode"] = null
+      constants["mobileNetworkCode"] = null
+    }
+    return constants
+  }
+
+  @ExpoMethod
+  fun requestPhoneStatePermissionsAsync(promise: Promise?) {
+    mPermissions.askForPermissionsWithPromise(promise, Manifest.permission.READ_PHONE_STATE)
+  }
+
+  @ExpoMethod
+  fun getPhoneStatePermissionsAsync(promise: Promise) {
+    mPermissions.getPermissionsWithPromise(promise, Manifest.permission.READ_PHONE_STATE)
+  }
+
+  @ExpoMethod
+  fun getCellularGenerationAsync(promise: Promise) {
+    try {
+      val networkType = getNetworkGeneration()
+      promise.resolve(networkType)
+    } catch (e: SecurityException) {
+      promise.reject("E_MISSING_PERMISSIONS", e)
+    } catch (e: Exception) {
+      promise.reject("ERR_CELLULAR_GENERATION_UNKNOWN_NETWORK_TYPE", "Unable to access network type or not connected to a cellular network", e)
+    }
+  }
+
+  @ExpoMethod
+  fun getCurrentCarrierAsync(promise: Promise) {
+    val carrierInfo = Bundle()
+    carrierInfo.putSerializable("allowsVoip", SipManager.isVoipSupported(mContext))
+    val telephonyManager = mContext.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
+    carrierInfo.putSerializable("isoCountryCode", telephonyManager.simCountryIso)
+
+    if (telephonyManager.simState == TelephonyManager.SIM_STATE_READY) {
+      carrierInfo.putSerializable("carrier", telephonyManager.simOperatorName)
+      carrierInfo.putSerializable("mobileCountryCode", telephonyManager.simOperator.take(3))
+      carrierInfo.putSerializable("mobileNetworkCode", StringBuilder(telephonyManager.simOperator).delete(0, 3).toString())
+      try {
+        carrierInfo.putSerializable("generation", getNetworkGeneration())
+      } catch (e: SecurityException) {
+        promise.reject("E_MISSING_PERMISSIONS", e)
+      }  catch (e: Exception) {
+        promise.reject("ERR_CELLULAR_GENERATION_UNKNOWN_NETWORK_TYPE", "Unable to access network type or not connected to a cellular network", e)
+      }
+    } else {
+      carrierInfo.putSerializable("carrier", null)
+      carrierInfo.putSerializable("mobileCountryCode", null)
+      carrierInfo.putSerializable("mobileNetworkCode", null)
+      carrierInfo.putSerializable("generation", null)
+    }
+    promise.resolve(carrierInfo)
+  }
+
+  @SuppressLint("MissingPermission")
+  private fun getNetworkGeneration(): Int {
+    if (checkPermissions()) {
```

Comment:
The early return reduces nesting and inlines the one line of code:
```suggestion
    if (!mPermissions.hasGrantedPermissions(Manifest.permission.READ_PHONE_STATE)) {
      return CellularGeneration.UNKNOWN.value
    }
```